### ISSUE TITLE
Bug: `thread '<unnamed>' panicked at 'Index 0 for dropped Node in NodeCollection'`

### ISSUE THREAD

**Issue by b-ma** [orottier/web-audio-api-rs#389]
I run this quite a few time `RUST_BACKTRACE=1 cargo run --example granular`

Got that:

```
thread '<unnamed>' panicked at 'Index 0 for dropped Node in NodeCollection', src/render/node_collection.rs:104:32
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: <web_audio_api::render::node_collection::NodeCollection as core::ops::index::IndexMut<web_audio_api::context::AudioNodeId>>::index_mut::{{closure}}
             at ./src/render/node_collection.rs:104:32
   3: core::option::Option<T>::unwrap_or_else
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/option.rs:979:21
   4: <web_audio_api::render::node_collection::NodeCollection as core::ops::index::IndexMut<web_audio_api::context::AudioNodeId>>::index_mut
             at ./src/render/node_collection.rs:100:9
   5: web_audio_api::render::graph::Graph::render
             at ./src/render/graph.rs:483:10
   6: web_audio_api::render::thread::RenderThread::render_inner
             at ./src/render/thread.rs:299:42
   7: web_audio_api::render::thread::RenderThread::render::{{closure}}
             at ./src/render/thread.rs:220:39
   8: no_denormals::no_denormals
             at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/no_denormals-0.1.2/src/lib.rs:102:12
   9: web_audio_api::render::thread::RenderThread::render
             at ./src/render/thread.rs:220:9
  10: web_audio_api::io::cpal::spawn_output_stream::{{closure}}
             at ./src/io/cpal.rs:481:17
  11: cpal::traits::DeviceTrait::build_output_stream::{{closure}}
             at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.2/src/traits.rs:169:17
  12: cpal::host::coreaudio::macos::Device::build_output_stream_raw::{{closure}}
             at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.2/src/host/coreaudio/macos/mod.rs:708:13
  13: coreaudio::audio_unit::render_callback::<impl coreaudio::audio_unit::AudioUnit>::set_render_callback::{{closure}}
             at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/coreaudio-rs-0.11.2/src/audio_unit/render_callback.rs:502:19
  14: coreaudio::audio_unit::render_callback::input_proc
             at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/coreaudio-rs-0.11.2/src/audio_unit/render_callback.rs:768:9
  15: _DefaultOutputAUFactory
  16: _AUNotQuiteSoSimpleTimeFactory
  17: _AUGenericOutputFactory
  18: <unknown>
  19: <unknown>
  20: <unknown>
  21: __pthread_start
```

**Comment by b-ma** [comment]
With `BACKTRACE=full`

```
thread '<unnamed>' panicked at 'Index 0 for dropped Node in NodeCollection', src/render/node_collection.rs:104:32
stack backtrace:
   0:        0x1016cec13 - std::backtrace_rs::backtrace::libunwind::trace::h14243dc752b5beac
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:        0x1016cec13 - std::backtrace_rs::backtrace::trace_unsynchronized::h71ecb3a9455536a6
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x1016cec13 - std::sys_common::backtrace::_print_fmt::h7e307467d6e88eaf
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:65:5
   3:        0x1016cec13 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he60192cdeae7ff98
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:44:22
   4:        0x1016ed73b - core::fmt::rt::Argument::fmt::hef24bdd6ecabdaa4
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/rt.rs:138:9
   5:        0x1016ed73b - core::fmt::write::hccf73e6ac9b102cd
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/fmt/mod.rs:1094:21
   6:        0x1016cc3de - std::io::Write::write_fmt::h2da296a2a97aa10d
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/io/mod.rs:1714:15
   7:        0x1016cea0a - std::sys_common::backtrace::_print::h57786acb50d0c08a
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:47:5
   8:        0x1016cea0a - std::sys_common::backtrace::print::h57a90601b74ca1dc
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:34:9
   9:        0x1016cfee5 - std::panicking::default_hook::{{closure}}::h8463146ca7fe4cbf
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:269:22
  10:        0x1016cfc8f - std::panicking::default_hook::h473c81bbe30d7761
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:288:9
  11:        0x1016d0440 - std::panicking::rust_panic_with_hook::h0ff95e3e20549b0d
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:705:13
  12:        0x1016d0363 - std::panicking::begin_panic_handler::{{closure}}::h5e967ae12a1d4ba1
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:597:13
  13:        0x1016cf019 - std::sys_common::backtrace::__rust_end_short_backtrace::h5c230463e113192d
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:151:18
  14:        0x1016d009d - rust_begin_unwind
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
  15:        0x101705dd3 - core::panicking::panic_fmt::h9d91d0e3e7a6cd1a
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
  16:        0x100c437ea - <web_audio_api::render::node_collection::NodeCollection as core::ops::index::IndexMut<web_audio_api::context::AudioNodeId>>::index_mut::{{closure}}::h11b2a3dfc317d774
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/node_collection.rs:104:32
  17:        0x100bdbdae - core::option::Option<T>::unwrap_or_else::h2807ff8b1390d856
                               at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/option.rs:979:21
  18:        0x100b78d4b - <web_audio_api::render::node_collection::NodeCollection as core::ops::index::IndexMut<web_audio_api::context::AudioNodeId>>::index_mut::h07969699e7e09ef4
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/node_collection.rs:100:9
  19:        0x100b78d4b - web_audio_api::render::graph::Graph::render::h9796256f8ea92d62
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/graph.rs:483:10
  20:        0x100c2c4e3 - web_audio_api::render::thread::RenderThread::render_inner::hdb0ba7fa7554deae
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/thread.rs:299:42
  21:        0x100c22847 - web_audio_api::render::thread::RenderThread::render::{{closure}}::h6281c7349933840d
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/thread.rs:220:39
  22:        0x100bc7f21 - no_denormals::no_denormals::h9016b2eaaf445690
                               at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/no_denormals-0.1.2/src/lib.rs:102:12
  23:        0x100c21afe - web_audio_api::render::thread::RenderThread::render::h29467cb3586205dd
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/render/thread.rs:220:9
  24:        0x100b7c1a5 - web_audio_api::io::cpal::spawn_output_stream::{{closure}}::hdcc2952c29e62af6
                               at /Users/me/work/dev/libs/web-audio-api/web-audio-api-rs/src/io/cpal.rs:481:17
  25:        0x100baa0c5 - cpal::traits::DeviceTrait::build_output_stream::{{closure}}::h78f888a0ca4a2e15
                               at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.2/src/traits.rs:169:17
  26:        0x100b8c823 - cpal::host::coreaudio::macos::Device::build_output_stream_raw::{{closure}}::hb18f3e397651d4a0
                               at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.2/src/host/coreaudio/macos/mod.rs:708:13
  27:        0x100bc4e28 - coreaudio::audio_unit::render_callback::<impl coreaudio::audio_unit::AudioUnit>::set_render_callback::{{closure}}::hd15f4722f814cf38
                               at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/coreaudio-rs-0.11.2/src/audio_unit/render_callback.rs:502:19
  28:        0x100f60496 - coreaudio::audio_unit::render_callback::input_proc::h9a331849b5aa6b47
                               at /Users/me/.cargo/registry/src/index.crates.io-6f17d22bba15001f/coreaudio-rs-0.11.2/src/audio_unit/render_callback.rs:768:9
  29:        0x1044da1be - _DefaultOutputAUFactory
  30:        0x1045e39c2 - _AUNotQuiteSoSimpleTimeFactory
  31:        0x1044de191 - _AUGenericOutputFactory
  32:     0x7ff81143250c - <unknown>
  33:     0x7ff811430041 - <unknown>
  34:     0x7ff81159f93e - <unknown>
  35:     0x7ff80f2991d3 - __pthread_start
```

**Comment by orottier** [comment]
Great that you managed to capture it. Will look into this with priority

**Comment by b-ma** [comment]
I'm a bit confused by this syntax, so maybe I'm mistaking, but wonder if that work as expected:

https://github.com/orottier/web-audio-api-rs/blob/ba7b5f36323fe349952b7d72615fd92e548db649/src/render/thread.rs#L273C1-L276C10
```rs
        if !self.graph.as_ref().is_some_and(Graph::is_active) {
            output_buffer.fill(S::from_sample_(0.));
            return;
        }
```

cf. [#338](https://github.com/orottier/web-audio-api-rs/pull/338)

**Comment by b-ma** [comment]
Ok I think I got it, the `NodeCollection` is actually never empty because `instance.ensure_capacity(64)` is called in constructor, maybe

```
let mut vec: Vec<Option<usize>> = vec![];
vec.resize_with(64, || None);
vec.is_empty()
> false
```

**Comment by b-ma** [comment]
Maybe the best thing is just to ensure `AudioDestination` exists

```rs
impl NodeCollection {
    pub fn is_empty(&self) -> bool {
        self.nodes[DESTINATION_NODE_ID.0].is_some()
    }
}
```

**Comment by orottier** [comment]
Fixed with #390
